### ISSUE TITLE
[Feature Request] Keep the finer granular error code for MiscellaneousError 

### ISSUE THREAD

**Issue by areshand** [move-language/move#66]
# 🚀 Feature Request

## Motivation

Multiple different errors all mapped to MiscellaneousError ex:
1. https://github.com/areshand/move/blob/0747a72af49e7116ae2b54625668d8536ee086d8/language/move-core/types/src/vm_status.rs#L182
2. https://github.com/areshand/move/blob/0747a72af49e7116ae2b54625668d8536ee086d8/language/move-core/types/src/vm_status.rs#L187

The impact is user lacks a clear idea of what went wrong when their txn failed. Ex:
when users published modules, both the **linker_error** and **MODULE_ADDRESS_DOES_NOT_MATCH_SENDER** are mapped to this MiscellaneousError, which doesn't help user to understand what went wrong. 

## Pitch

**Describe the solution you'd like**
Keep the finer granular status code when returning the MiscellaneousError to provide more detailed error message to the users

**Comment by tnowacki** [comment]
Pulling from @sblackshear's detailed answer on discord (Thanks Sam!)

"For some historical context, I believe we used to support a much broader set of error codes, but were having difficulty with error codes committing to implementation details of the Move VM or the adapter. The current error behavior is a reaction to this experience. Error behavior is part of the VM spec, so it's important to keep it as simple/minimal as possible. That is (unfortunately) at odds with devex, but as Todd says, the intention is that this belongs in tooling outside the VM or in a layer above like the adapter"

In short, this isn't the purpose of VMStatus, as an adapter, you can use VMError as much as you want, and report this back to the developer as you see fit. The VM doesn't guarantee being stable enough to always return the same type checking error. It guarantees Yes/No answers on verification. And then guarantees the position for any Move Abort

**Comment by tnowacki** [comment]
Closing for now as it won't be something that is changed. But happy to help with sorting out any information that is missing in `VMError` or for strategies in logging that error

**Comment by areshand** [comment]
thanks for the detailed answer.
